### PR TITLE
Creates the RandomAugmentationPipeline Layer

### PR DESCRIPTION
**Pull request by LukeWood**
Example from the demo:

Figure_1

### CODE REVIEW

**Comment by LukeWood** [comment]
Interested in your feedback on the RandomAugmentationPipeline class.  I feel it could have a better name.

also: I’m not sure what’s wrong with the unit tests.  Perhaps some weird caching due to the map function is going on?

**Comment by innat** [comment]
What are the uses of this layer `RandomAugmentationPipeline`? Is it general for all KPL or specific for `RandAugment`? Does it mean that the `RandomAugmentationPipeline` will take randomly apply `layers` in a different order?

**Comment by LukeWood** [comment]
> What are the uses of this layer `RandomAugmentationPipeline`? Is it general for all KPL or specific for `RandAugment`? Does it mean that the `RandomAugmentationPipeline` will take randomly apply `layers` in a different order?

Yeah, RandAugment refers to a specific preprocessing layer.  RandomAugmentationPipeline is an abstraction to randomly select from a list of transformations `augmentations_per_image` times.  I don't love the name RandomAugmentationPipeline, but couldn't think of anything more descriptive

**Comment by bhack** [comment]
> I don't love the name RandomAugmentationPipeline, but couldn't think of anything more descriptive

As this could be more complex that just a random sampling from a list we could think at this as RandomPolicy over a list of augmentation layers.

**Comment by LukeWood** [comment]
> RandomPolicy

RandomPolicy is nice.  RandomlySampledPolicy maybe?

**Comment by LukeWood** [comment]
CC @mattdangerw because he may have some good naming advice

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    References:
+        - [RandAugment](https://arxiv.org/abs/1909.13719)
+
+    Args:
+        value_range: the range of values the incoming images will have.
+            Represented as a two number tuple written [low, high].
+            This is typically either `[0, 1]` or `[0, 255]` depending
+            on how your preprocessing pipeline is setup.
+        augmentations_per_image: the number of layers to use in the rand augment policy.
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:46]: Default to 3.

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        magnitude: the magnitude to use for each of the augmentation.  magnitude should
+            be a float in the range `[0, 1]`.  A magnitude of `0` indicates that the
+            augmentations are as weak as possible (not recommended), while a value of
+            `1.0` implies use of the strongest possible augmentation.  Defaults to
+            `0.5`.
+        magnitude_stddev: the standard deviation to use when drawing values
+            for the perturbations.  Keep in mind magnitude will still be clipped to the
+            range `[0, 1]` after samples are drawn from the uniform distribution.
+            Defaults to `0.15`.
+        rate:  the rate at which to apply each augmentation.  This parameter is applied
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:56]: range from [0, 1] and default to 1.

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+            on a per-distortion layer, per image.
+
+    Usage:
+    ```python
+    (x_test, y_test), _ = tf.keras.datasets.cifar10.load_data()
+    rand_augment = keras_cv.layers.RandAugment(
+        value_range=(0, 255), augmentations_per_image=3, magnitude=5.0
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:63]: magnitude should be 0 to 1, rather than 5.0

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+            ),
+            augmentations_per_image=augmentations_per_image,
+            rate=rate,
+            **kwargs,
+            seed=seed,
+        )
+        self.magnitude = float(magnitude)
+        self.value_range = value_range
+        self.seed = seed
+        if magnitude < 0.0 or magnitude > 1:
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:93]: Should we do the validation check before calling any init? (just raise early and be efficient.)

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            raise ValueError(
+                f"`magnitude` must be in the range [0, 1], got `magnitude={magnitude}`"
+            )
+        self.magnitude_stddev = float(magnitude_stddev)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:97]: Should we also do any validation for param like magnitude_stddev as well?

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                "value_range": self.value_range,
+                "augmentations_per_image": self.augmentations_per_image,
+                "magnitude": self.magnitude,
+                "magnitude_stddev": self.magnitude_stddev,
+                "rate": self.rate,
+                "seed": self.seed,
+            }
+        )
+        # layers is recreated in the constructor
+        del config["layers"]
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/rand_augment.py:173]: Are the layers created from parent class?

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    # create the pipeline.
+    pipeline = keras_cv.layers.RandomAugmentationPipeline(
+        layers=layers, augmentations_per_image=3
+    )
+
+    augmented_images = pipeline(images)
+    ```
+
+    Args:
+        layers: a list of `keras.Layers`.  These are randomly sampled during
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_augmentation_pipeline.py:49]: Is there any particular requirement for the layer? eg it need to be BaseImageAugmentationLayer?

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs, seed=seed)
+        self.augmentations_per_image = augmentations_per_image
+        self.rate = rate
+        self.layers = layers
+        self.auto_vectorize = auto_vectorize
+        self.seed = seed
+
+    def _augment(self, sample):
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_augmentation_pipeline.py:79]: In the parent layer, do we call it sample or inputs? (just want to be consistent here.)

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            default=lambda: sample,
+        )
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "augmentations_per_image": self.augmentations_per_image,
+                "rate": self.rate,
+                "layers": self.seed,
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_augmentation_pipeline.py:112]: shouldn't this be "seed": self.seed?

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        if skip_augment > self.rate:
+            return sample
+
+        selected_op = self._random_generator.random_uniform(
+            (), minval=0, maxval=len(self.layers), dtype=tf.int32
+        )
+
+        branch_fns = []
+        for (i, layer) in enumerate(self.layers):
+            branch_fns.append((i, lambda: layer(sample)))
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_augmentation_pipeline.py:98]: we will need to be careful for this, eg define the lambda in a for loop. I have seen issues about this where lambda only capture the first invocation. See https://github.com/keras-team/keras/issues/16316#event-6390388564

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    ):
+        super().__init__(**kwargs, seed=seed)
+        self.augmentations_per_image = augmentations_per_image
+        self.rate = rate
+        self.layers = layers
+        self.auto_vectorize = auto_vectorize
+        self.seed = seed
+
+    def _augment(self, sample):
+        result = sample.copy()
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_augmentation_pipeline.py:80]: Any reason we need to do copy here?

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 7 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py:49]: I think this is nicer and decouple the sublayers with the container layer.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 1 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/rand_augment.py:46]: Can this be a range?

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:63]: Thanks - updated

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:93]: updated to be first line of __init__

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:97]: Done.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 8 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py:79]: Updated to _inputs.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 9 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py:112]: Good catch - surprised this worked.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 11 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py:80]: Updated to remove copy().

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:46]: Done.

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:56]: Done.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py:49]: Added that these must subclass.  Originally I didn't want to point users this way,  but otherwise we'd have to show them how to augmented the dictionary inputs.  We should probably expose BaseImageAugmentationLayer someday, but not yet.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 10 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py]: If the unit tests pass are you concerned about this?  I think this will be fine - but maybe I am missing something.

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:181]: Yeah, if we don't del this RandAugment serialization fails as it tries to provide `layers` twice in the constructor.

File: keras_cv/layers/preprocessing/random_augmentation_pipeline.py (hunk 10 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_augmentation_pipeline.py]: I was bit worried that the current unit test didn't catch this since all the layers/lambda are just the same. But maybe we are fine, since the sample here didn't change across the loop.

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+        - [RandAugment](https://arxiv.org/abs/1909.13719)
+
+    Args:
+        value_range: the range of values the incoming images will have.
+            Represented as a two number tuple written [low, high].
+            This is typically either `[0, 1]` or `[0, 255]` depending
+            on how your preprocessing pipeline is setup.
+        augmentations_per_image: the number of layers to use in the rand augment policy.
+            Defaults to 3.
+        magnitude: the magnitude to use for each of the augmentation.  magnitude should
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/rand_augment.py:48]: We should clarify in the description that the actual magnitude is random and that this is the mean (before truncation to [0, 1])

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+            be a float in the range `[0, 1]`.  A magnitude of `0` indicates that the
+            augmentations are as weak as possible (not recommended), while a value of
+            `1.0` implies use of the strongest possible augmentation.  Defaults to
+            `0.5`.
+        magnitude_stddev: the standard deviation to use when drawing values
+            for the perturbations.  Keep in mind magnitude will still be clipped to the
+            range `[0, 1]` after samples are drawn from the uniform distribution.
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/rand_augment.py:55]: uniform -> normal

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+            Defaults to `0.15`.
+        rate:  the rate at which to apply each augmentation.  This parameter is applied
+            on a per-distortion layer, per image.  Should be in the range [0, 1].
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/rand_augment.py:58]: Nit: use backticks.

A related question is whether numbers (e.g. `0.15`) should use backticks or not -- here you do it for floats but not for ints. Very minor question though.

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+class RandAugment(RandomAugmentationPipeline):
+    """RandAugment performs the Rand Augment operation on input images.
+
+    This layer can be thought of as an all in one image augmentation layer.  The policy
+    implemented by this layer has been benchmarked extensively and is effective on a
+    wide variety of datasets.
+
+    The policy operates as follows:
+
+    For each `layer`, the policy selects a random operation from a list of operations.
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/rand_augment.py:34]: "Layer" no longer refers to an actual signature argument, so the backticks are confusing -- prefer something like "For each augmentation in the range `[0, augmentations_per_image]`, "

File: keras_cv/layers/preprocessing/rand_augment.py [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+
+    def __init__(
+        self,
+        value_range,
+        augmentations_per_image=3,
+        magnitude=0.5,
+        magnitude_stddev=0.15,
+        rate=1.0,
+        seed=None,
+        **kwargs,
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/rand_augment.py:79]: Signature LGTM

**Review by fchollet — COMMENTED** [review]
Thanks for the update -- it looks very nice now!

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 1 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/rand_augment.py:46]: Done? What?

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 15 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/rand_augment.py:34]: Can we sample in this range? Or do we need to always execute all the ops in a fixed range?

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 12 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py]: Updated docstring

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 14 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:59]: Updated

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 15 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:34]: @bhack Rate will handle this need.

Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py]: Updated docstring

File: keras_cv/layers/preprocessing/rand_augment.py (hunk 16 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/rand_augment.py:81]: Thanks

**Review by fchollet — APPROVED** [review]
LGTM, thanks!
